### ISSUE TITLE
ParseWithClaims: getting json error when processing JWT that has multiple values for audience

### ISSUE THREAD

**Issue by jeffkayser2** [dgrijalva/jwt-go#348]
I am using JWT tokens generated by Auth0 and oauth2.  The token payload looks like this:

`{
  "iss": "https://redacted.auth0.com/",
  "sub": "google-oauth2|redacted",
  "aud": [
    "https://redacted:443/app/v1",
    "https://redacted.auth0.com/userinfo"
  ],
  "iat": 1565196366,
  "exp": 1565282766,
  "azp": "redacted",
  "scope": "openid profile email",
  "permissions": [
    "admin"
  ]
}`

Because the audience is multi-valued, I am getting parsing error when trying to read the claims.

My parsing code looks like this:

`type CustomClaimType struct {
        Permissions     []string        `json:"permissions"`
        Scope           string          `json:"scope"`
        jwt.StandardClaims
}`

. . .

`// Now parse the token
parsedToken, err := jwt.ParseWithClaims(bearerToken, &CustomClaimType{}, auth0ValidationKeyGetter)
if err != nil {
        log.Printf("Got error parsing token: " + err.Error())
        return nil, false, err
}`

When I run this, I get an error:

Got error parsing token: json: cannot unmarshal array into Go struct field CustomClaimType.aud of type string

**Comment by jeffkayser2** [comment]
[Issue348_Fix.zip](https://github.com/dgrijalva/jwt-go/files/3478283/Issue348_Fix.zip)

**Comment by jeffkayser2** [comment]
I have fixed the code, but ran into an issue trying to push the change to github, so I attached a ZIP file of the changes instead.

**Comment by jugaltheshah** [comment]
Having the same issue. Wouldn't this fix cause backwards compatibility issues @jeffkayser2? I'm very new to Go so I don't know if this is possible, but it almost seems Audience needs to be defined as `Array | string`, which also matches the spec: https://tools.ietf.org/html/rfc7519#section-4.1.3

**Comment by jeffkayser2** [comment]
Good point.  Could make it an interface{} instead.  Would have to work through the various test cases to ensure that both Array and string claims are properly decoded and validated.

**Comment by dklesev** [comment]
@jeffkayser2 why you don't make a PR?

**Comment by dklesev** [comment]
Refactored in favour of your solution (I think it's cleaner, hope it's ok for you to push parts of your code), please see #355

**Comment by jeffkayser2** [comment]
No problem! Thanks for doing that.  One of these days, I'll figure out git.  :-)

**Comment by friedrich** [comment]
Is this going to be fixed? Currently, RFC compliant tokens cannot be parsed because of this problem, which is a major issue in my opinion. Is there anything wrong with the proposed pull requests?

**Comment by coilnovasince1985** [comment]
Until now issue is not resolved.

**Comment by mileusna** [comment]
I took a sneak peek into v4.0.0 preview source code (https://github.com/dgrijalva/jwt-go/releases) and it looks like that v4 will fix this, but I don't know when it will be released.

**Comment by zbindenren** [comment]
I can confirm v4.0.0 preview solves this issue.

**Comment by chakch** [comment]
Any stable release to fix this issue ?

**Comment by ivucica** [comment]
@dgrijalva Could you eventually put the contents of `v4.0.0-preview` into a subfolder named `v4`, so that neither the adoption of Go modules nor use of vendoring are required?

**Comment by trfv** [comment]
[note]
you can use golang-jwt instead of jwt-go.

https://github.com/golang-jwt/jwt
